### ISSUE TITLE
Future not resolving with Tonic and SimpleSpanProcessor

### ISSUE THREAD

**Issue by duarten** [open-telemetry/opentelemetry-rust#785]
Consider the following otlp configuration:

```rust
opentelemetry_otlp::new_exporter()
        .tonic()
        .with_endpoint(endpoint.as_str())
        .with_metadata(metadata)
        .with_tls_config(
            tonic::transport::ClientTlsConfig::new().domain_name(endpoint.host_str().unwrap()),
        )
        .with_protocol(opentelemetry_otlp::Protocol::Grpc)
        .with_timeout(std::time::Duration::from_secs(3))
```

When using `SimpleSpanProcessor` (through `with_simple_exporter`), eventually the program reaches a point where the future actually exporting the span never resolves. It's [this line](https://github.com/open-telemetry/opentelemetry-rust/blob/main/opentelemetry-sdk/src/trace/span_processor.rs#L107): the thread dequeues the span, and `block_on` never returns. It seems this is triggering a deadlock.

I'm using `tracing-rs` (with the workaround in #473) on top and  `opentelemetry-otlp` configured with `["tonic", "tls", "tls-roots"] `.

Curiously, if I use the following very simple processor, everything works correctly:

```rust
#[derive(Debug)]
struct SimplerProcessor<E: SpanExporter + Send + Sync> {
    inner: std::sync::Arc<std::sync::Mutex<E>>,
}

impl<E: SpanExporter + Sync + Send> SpanProcessor for SimplerProcessor<E> {
    fn on_start(&self, _span: &mut Span, _cx: &opentelemetry::Context) {
    }

    fn on_end(&self, span: SpanData) {
        tokio::task::block_in_place(|| {
            tokio::runtime::Handle::current()
                .block_on(self.inner.lock().unwrap().export(vec![span]))
        })
        .expect("span not sent");
    }

    fn force_flush(&self) -> opentelemetry::trace::TraceResult<()> {
        Ok(())
    }

   fn shutdown(&mut self) -> opentelemetry::trace::TraceResult<()> {
        Ok(())
    }
}
```

Does any of this ring any bells?

**Comment by TommyCpp** [comment]
IIRC `SimpleSpanProcessor` is not compatilbe with async runtimes. The tricky part here is we want to support different runtime so we generally need a `Runtime` to know how to spawn a task.

**Comment by duarten** [comment]
Oh! I was getting the same issue with the batch processor, but I didn't debug it thoroughly. I think it also calls `futures_executor::block_on` at some point and that may cause the issue, but I'll look more deeply.

Did you consider providing an `AsyncRuntime`? `rdkafka` does something [like that](https://github.com/fede1024/rust-rdkafka/blob/9594a34f73709d443c968ab191a010eb1908dae0/src/util.rs#L351) (it is meant to be implemented by empty structs).

**Comment by TommyCpp** [comment]
Yeah we do have [Runtime](https://docs.rs/opentelemetry/latest/opentelemetry/runtime/trait.Runtime.html) and it's one of the parameters of the BatchSpanProcessor.

**Comment by duarten** [comment]
Oh, nice. Indeed I'm having better luck with the `BatchSpanProcessor`. I have to call `force_flush_tracer_provider` since I'm running in a Lambda, and that still calls `futures_executor::block_on`. I imagine it would deadlock if I tried to run this on a single threaded runtime. It might be a good idea to lift the `block_on` to the `Runtime` trait.

**Comment by TommyCpp** [comment]
Yep we have encounter it before if user are using tokio single threaded runtime. [TokioCurrentThread](https://docs.rs/opentelemetry/latest/opentelemetry/runtime/struct.TokioCurrentThread.html#) runtime, [which will spawn the resource on a different thread](https://docs.rs/opentelemetry/latest/src/opentelemetry/runtime.rs.html#87-92). 

This is not a perfect design but `force_flush` need at least two thread(or a runtime to allow at least two task running in parallel). The reason is we need to block on the main thread so that it will not return while the background thread finish the work.

I think if we can make `force_flush` an async function/returns a furture things might be easier. But we do want to make sure the API can work without an async runtime.
